Disable (Broken) CUDAStreamVariable in dynamo

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100766

While attempting to explore XLTransformers w/ PT2, I found that we leak tracing time objects (VariableTrackers) into the runtime:

```
Traceback (most recent call last):
  File "/scratch/voz/work/xlformers/train.py", line 686, in <module>
    main(cfg)
  File "/scratch/voz/work/xlformers/train.py", line 357, in main
    pred, _ = model(x)
  File "/scratch/voz/work/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/scratch/voz/work/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch/voz/work/pytorch/torch/_dynamo/eval_frame.py", line 282, in _fn
    return fn(*args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1416, in forward
    self._lazy_init()
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1424, in <resume in forward>
    args, kwargs = cast_floats_to_right_precision(True, True, *args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1434, in <resume in forward>
    self._rebuild_full_params()
  File "/scratch/voz/work/pytorch/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1932, in _rebuild_full_params
    def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1932, in <resume in _rebuild_full_params>
    def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
  File "/scratch/voz/work/pytorch/torch/cuda/__init__.py", line 464, in __enter__
    if self.src_prev_stream.device != cur_stream.device:
AttributeError: 'CUDAStreamVariable' object has no attribute 'device'
```

This indicates a serious bug. 

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Disable CUDAStreamVariable in dynamo

[ghstack-poisoned]

wconstab commented:
note that we don't have any handling for fairscale FSDP in torchdynamo.  Don't confuse it with pytorch-distributed-FSDP, which is a different codebase.

It is probably a good idea to first identify how to get dynamo to graph-break on fs-fsdp, but its also possible that's not enough

ezyang commented:
Test will be good when you get it

yanboliang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -503,12 +503,13 @@ def index_source(key):
+            #     value,
+            #     source=self.source,
+            #     guards=self.make_guards(GuardBuilder.ID_MATCH),
+            # )
```
I think we should identify what caused the error, it seems we missed ```device``` for ```CUDAStreamVariable```, can we add it? If you are urgent to unblock anything but don't want to graph break, you can also try to wrap it as null context manager and print warnings like this:
https://github.com/pytorch/pytorch/blob/970c60b336b7bc346cd1cbb8b86f547983c9668a/torch/_dynamo/variables/torch.py#L382-L389

wconstab replied:
Yanbo, would your suggestion to add device to CudaStreamVariable just paper over the fact that we somehow leaked a VariableTracker object outside of our compiler?  It's important to not let that happen.

voznesenskym committed:
Update on "Disable (Broken) CUDAStreamVariable in dynamo"


While attempting to explore XLTransformers w/ PT2, I found that we leak tracing time objects (VariableTrackers) into the runtime:

```
Traceback (most recent call last):
  File "/scratch/voz/work/xlformers/train.py", line 686, in <module>
    main(cfg)
  File "/scratch/voz/work/xlformers/train.py", line 357, in main
    pred, _ = model(x)
  File "/scratch/voz/work/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/scratch/voz/work/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch/voz/work/pytorch/torch/_dynamo/eval_frame.py", line 282, in _fn
    return fn(*args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1416, in forward
    self._lazy_init()
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1424, in <resume in forward>
    args, kwargs = cast_floats_to_right_precision(True, True, *args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1434, in <resume in forward>
    self._rebuild_full_params()
  File "/scratch/voz/work/pytorch/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1932, in _rebuild_full_params
    def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1932, in <resume in _rebuild_full_params>
    def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
  File "/scratch/voz/work/pytorch/torch/cuda/__init__.py", line 464, in __enter__
    if self.src_prev_stream.device != cur_stream.device:
AttributeError: 'CUDAStreamVariable' object has no attribute 'device'
```

This indicates a serious bug. 

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
Perf run here, small change but I dislike adding graph breaks without a perf run. https://github.com/pytorch/pytorch/actions/runs/4982837918

voznesenskym commented:
No perf impact

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot merge -r

voznesenskym committed:
Update on "Disable (Broken) CUDAStreamVariable in dynamo"


While attempting to explore XLTransformers w/ PT2, I found that we leak tracing time objects (VariableTrackers) into the runtime:

```
Traceback (most recent call last):
  File "/scratch/voz/work/xlformers/train.py", line 686, in <module>
    main(cfg)
  File "/scratch/voz/work/xlformers/train.py", line 357, in main
    pred, _ = model(x)
  File "/scratch/voz/work/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/scratch/voz/work/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch/voz/work/pytorch/torch/_dynamo/eval_frame.py", line 282, in _fn
    return fn(*args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1416, in forward
    self._lazy_init()
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1424, in <resume in forward>
    args, kwargs = cast_floats_to_right_precision(True, True, *args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1434, in <resume in forward>
    self._rebuild_full_params()
  File "/scratch/voz/work/pytorch/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1932, in _rebuild_full_params
    def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
  File "/data/home/voz/miniconda3/envs/torch5/lib/python3.10/site-packages/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 1932, in <resume in _rebuild_full_params>
    def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
  File "/scratch/voz/work/pytorch/torch/cuda/__init__.py", line 464, in __enter__
    if self.src_prev_stream.device != cur_stream.device:
AttributeError: 'CUDAStreamVariable' object has no attribute 'device'
```

This indicates a serious bug. 

cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
